shared/editor/plugins/HeadingTracker.ts

```diff
@@ -0,0 +1,224 @@
+ *
+ * Example:
+ *
+ *    Consider the following document structure,
+ *     _____________________________________________________________________________
+ *    | H2(c)                                                                       |
+ *    |   This goes under H2                                                        |
+ *    |                                                                             |
+ *    | H1(c)                                                                       |
+ *    |   This goes under H1                                                        |
+ *    |   H3(c)                                                                     |
+ *    |     This goes under H3, which is under H1!                                  |
+ *    |   H3'(u)                                                                    |
+ *    |     This goes under H3', which is under H1!                                 |
+ *    |     H4(c)                                                                   |
+ *    |       This goes under H4, which is under H3, which, further, is under H1!   |
+ *    |_____________________________________________________________________________|
+ *
+ *    The (c) & (u) denote folded and unfolded states of the heading, respectively.
+ *
+ *    Now, let's say the cursor lands under H4 upon backspacing. The desired state, after backspacing should be that H1,
+ *    along with its "descendant" H4 end up being unfolded, displaying respective content under them.
+ *
+ *    So, Let's look at the the closest heading of each level from the cursor,
+ *    Level 1: H1
+ *    Level 2: H2
+ *    Level 3: H3'(and not H3)
+ *    Level 4: H4
+ *
+ *    And, the stack maintaing the invariant would look like,
+ *
+ *    | H4 |
+ *    | -- |
+ *    | H3'|
+ *    | -- |
+ *    | H1 |
+ *     ----
+ *
+ *    * Level of H4(4) > Level of H3'(3) && H4 is closer to the cursor than H3'
+ *    * Level of H3'(3) > Level of H1(1) && H3' is closer to the cursor than H1
+ *
+ *    Notice that we didn't push H2 onto the stack after H1, because that would violate the invariant considering
+ *    H2 is further away from the cursor than H1, and therefore, is NOT A DESCENDANT of H1.
+ *    So, we don't care about H2's collapsed state in this particular case.
+ *
+ *    Now, we pop out headings and collect the ones which are collapsed, i.e, H1 and H4. Then, we unfold them.
+ */
+
+export class HeadingTracker extends Plugin {
+  private sortedHeadingPositions: SortedHeadingPositions;
+  private closestHeadingsStack: Stack<HeadingWithPos>;
+  private doc: Node;
+
+  constructor() {
+    super({
+      appendTransaction: (transactions, _oldState, newState) => {
+        if (!transactions.some((tr) => tr.docChanged)) {
+          return;
+        }
+        this.doc = newState.doc;
+        this.updateHeadingPositions();
+        const cursorPos = newState.selection.from;
+        this.collectPrevClosestHeadingsFrom(cursorPos);
```

Comment:
Some after-thoughts on this...
Just pointing it out that this runs upon _every_ transaction that triggers a doc update. And, as a result of that it build up the entire sorted list containing heading positions in the doc, every time! Kind of inefficient perf-wise but hopefully works _fine_ out in the wild for most cases. That being said, it'd be great to figure out a way to just update the list(and not rebuild everytime), the way decorations are mapped through changes.

Reply:
Yea, this is a lot for a very small fix – I think if we're going to do this we should at least combine the other logic which tracks headings for the table of contents and have it use this same underlying data structure?

Reply:
If the same data structure _can_ be used for another purpose, why not! I'm in favor of tackling this problem separately though.

Reply:
Okay, so there are two issues as I read it:
1. Mimicking `DecorationSet.map` to _our_ data structure so that it updates, rather than re-builds
2. Use the same data structure to _update_ headings in ToC

I'll try to tackle (2) above to re-use the data structure barring the perf-issue.